posthog/src/main/java/com/posthog/PostHog.kt

```diff
@@ -0,0 +1,567 @@
+    }
+
+    private var anonymousId: String
+        get() {
+            var anonymousId: String?
+            synchronized(anonymousLock) {
+                anonymousId = config?.cachePreferences?.getValue("anonymousId") as? String
+                if (anonymousId == null) {
+                    anonymousId = UUID.randomUUID().toString()
+                    this.anonymousId = anonymousId ?: ""
+                }
+            }
+            return anonymousId ?: ""
+        }
+        set(value) {
+            config?.cachePreferences?.setValue("anonymousId", value)
+        }
+
+    private var distinctId: String
+        get() {
+            return config?.cachePreferences?.getValue("distinctId", defaultValue = anonymousId) as? String ?: ""
+        }
+        set(value) {
+            config?.cachePreferences?.setValue("distinctId", value)
+        }
+
+    private fun buildProperties(
+        distinctId: String,
+        properties: Map<String, Any>?,
+        userProperties: Map<String, Any>?,
+        userPropertiesSetOnce: Map<String, Any>?,
+        groupProperties: Map<String, Any>?,
+        appendSharedProps: Boolean = true,
+    ): Map<String, Any> {
+        val props = mutableMapOf<String, Any>()
+
+        properties?.let {
+            props.putAll(it)
+        }
+
+        if (appendSharedProps) {
+            config?.context?.getStaticContext()?.let {
+                props.putAll(it)
+            }
+
+            config?.context?.getDynamicContext()?.let {
+                props.putAll(it)
+            }
+
+            if (config?.sendFeatureFlagEvent == true) {
+                featureFlags?.getFeatureFlags()?.let {
+                    if (it.isNotEmpty()) {
+                        val keys = mutableListOf<String>()
+                        for (entry in it.entries) {
+                            props["\$feature/${entry.key}"] = entry.value
+                            keys.add(entry.key)
+                        }
+                        props["\$active_feature_flags"] = keys
+                    }
+                }
+            }
+
+            val registeredPrefs = memoryPreferences.getAll()
+            if (registeredPrefs.isNotEmpty()) {
+                props.putAll(registeredPrefs)
+            }
+        }
+
+        userProperties?.let {
+            props["\$set"] = it
+        }
+
+        userPropertiesSetOnce?.let {
+            props["\$set_once"] = it
+        }
+
+        groupProperties?.let {
+            props["\$groups"] = it
+        }
+
+        // only set if not there.
+        props["distinct_id"]?.let {
+            props["distinct_id"] = distinctId
+        }
+
+        return props
+    }
+
+    // test: $merge_dangerously
+    public override fun capture(
+        event: String,
+        distinctId: String?,
+        properties: Map<String, Any>?,
+        userProperties: Map<String, Any>?,
+        userPropertiesSetOnce: Map<String, Any>?,
+        groupProperties: Map<String, Any>?,
```

Comment:
I wonder if it'd be better to have a data class here that would have an optional ctor param for each property, so the method param. list is smaller, I am fine either way